Adding content for deployment checklists (#525)
minio/docs: source/operations/checklists/hardware.rst

   :local:
   :depth: 2

Use the following checklist when planning the hardware configuration for a production, distributed MinIO deployment.

Considerations
--------------

When selecting hardware for your MinIO implementation, take into account the following factors:

- Expected amount of data in tebibytes to store at launch
- Expected growth in size of data for at least the next two years
- Number of objects by average object size
- Average retention time of data in years
- Number of sites to be deployed

Hardware Requirements and Recommendations
-----------------------------------------

.. note:: 

   For specific hardware suggestions, refer to https://min.io/product/reference-hardware.

.. list-table::
   :widths: auto
   :width: 100%

   * - :octicon:`circle`

>      - |Sufficient CPU cores to acheive performance goals

Author: Added some additional info to these two lines.

       |MinIO recommends Dual Intel® Xeon® Scalable Gold CPUs (minimum 8 cores per socket)

   * - :octicon:`circle`

>      - |Sufficient RAM to acheive performance goals

Author: Added additional info and links to the charts and formula

       |MinIO recommends at least 128 GB of RAM 

   * - :octicon:`circle`
     - Four nodes or servers

   * - :octicon:`circle`
     - |SATA/SAS drives for capacity and NVMe SSDs for high-performance
       |MinIO recommends a minimum of 8 drives per server

   * - :octicon:`circle`
     - |25GbE network for capacity 
       |100GbE Network interface cards for high performance

.. important:: 

>    Object storage operations are primarily throughput bound. 

Author: I expanded the admonition to include the three examples you gave here and additional clarification on network throughput, storage controller, and the storage devices.

   :local:
   :depth: 2

Use the following checklist when planning the hardware configuration for a production, distributed MinIO deployment.

Considerations
--------------

When selecting hardware for your MinIO implementation, take into account the following factors:

- Expected amount of data in tebibytes to store at launch
- Expected growth in size of data for at least the next two years
- Number of objects by average object size
- Average retention time of data in years
- Number of sites to be deployed

Hardware Requirements and Recommendations
-----------------------------------------

.. note:: 

>    For specific hardware suggestions, refer to https://min.io/product/reference-hardware.

Reviewer: We get a lot of confusion on whether or not we sell hardware, so maybe:

```suggestion
   See our `Reference Hardware <https://min.io/product/reference-hardware?ref=docs` page for a curated selection of servers and storage components from our hardware partners.
   
   MinIO does not provide any hosted services or hardware sales.

```

Author: I like it! Added.

.. list-table::
   :widths: auto
   :width: 100%

   * - :octicon:`circle`
     - | Sufficient CPU cores to acheive performance goals
       | MinIO recommends Dual Intel® Xeon® Scalable Gold CPUs (minimum 8 cores per socket)

   * - :octicon:`circle`
     - | Sufficient RAM to acheive performance goals
       | MinIO recommends at least 128 GB of RAM 

   * - :octicon:`circle`
     - Four nodes or servers

   * - :octicon:`circle`
     - | SATA/SAS drives for capacity and NVMe SSDs for high-performance
       | MinIO recommends a minimum of 8 drives per server

   * - :octicon:`circle`
     - | 25GbE network for capacity 
       | 100GbE Network interface cards for high performance

.. important:: 

   Object storage operations are primarily throughput bound. 
   MinIO takes full advantage of the modern hardware improvements such as AVX-512 SIMD acceleration, 100GbE networking, and NVMe SSDs when available.

> Hardware Tests to Complete

Reviewer: Maybe "Recommended Hardware Tests?

Author: 👍

--------------------------

MinIO Diagnostics
~~~~~~~~~~~~~~~~~

Run the following command.
If you have access to :ref:`SUBNET <minio-docs-subnet>`, you can upload the results there.

.. code-block:: shell
   :class: copyable

   mc support diag ALIAS

Drive Tests
~~~~~~~~~~~

> Run these tests outside of MinIO to verify capabilities of your hardware.

Reviewer: ```suggestion
Run these tests outside of MinIO to verify capabilities of your storage hardware.
```

Reviewer: Also maybe lead with the second graf:

"These tests <do a thing>.
Use the results to verify the capabilities..."

Probably don't need to say they are outside of MinIO.

Reviewer: Maybe just state that these are OS-level tools with no dependencies on MinIO.

Author: Done

Test each drive independently on all servers to ensure they are identical in performance.
Record the results for later reference.

#. Uncached write test

>    

Reviewer: We should try to say what this is doing. at a glance, it writes about 80K files of random data using ODIRECT to avoid any drive caching. If possible we should try to provide some output as guidance on what the results mean - @eco-minio might be able to help here. There are specific meaningful sections in the output.

Also instructions on how to clean up afterwards. Not sure if we can use `of=/dev/null` ?

Author: Created tables for each command to list what each part of the command is doing. Also added what to expect for the output, without giving a specific example.

    .. code-block::
       :class: copyable

       dd if=/dev/zero of=/mnt/driveN/testfile bs=128k count=80000 oflag=direct conv=fdatasync > dd-write-drive1.txt

> #. Uncached read test

Reviewer: I guess we should note that this depends on the previous test, and scratch the cleanup instructions until the end.

Same thought on providing some guidance.

Author: 👍

    .. code-block::
       :class: copyable

       dd if=/mnt/diskN/testfile of=/dev/null bs=128k iflag=direct > dd-read-drive1.txt

   Replace ``diskN`` with the path for the disk you are testing.

>    If you need to create a sufficiently sized file, you can use the ``head`` command.

Reviewer: We should state _why_ this is necessary.

IIRC there are both "small files" (<1MB) and "large files" (everything else) and "REALLY large files" in the 1GB+ range.

Not sure whether we can provide hard guidance here.

Author: Changed this line to the following:
```
Use a sufficiently sized file that mimics the primary use case for your deployment to get accurate read test results.
   
   Small files have a size < 1MB.
   Large files are from 1MB — 1GB.
   Really large files are > 1GB.

   You can use the ``head`` command to create a file to use.
   The following command example creates a 10 Gigabyte file called ``testfile``.
```

   The following command example creates a 10 Gigabyte file called ``testfile``.

    .. code-block:: shell
       :class: copyable

       head -c 10G </dev/urandom > testfile

#. IO Controller test
   

>    Use `IOzone <http://iozone.org/>`__ to test the input/output controllwer and all drives in combination.

Reviewer: ```suggestion
   Use `IOzone <http://iozone.org/>`__ to test the input/output controller and all drives in combination.
```

Author: De-Elmer Fudd-ed.

   Document the performance numbers for each server in your deployment.

   .. code-block:: shell
      :class: copyable

      iozone -s 1g -r 4m -i 0 -i 1 -i 2 -I -t 160 -F /mnt/sdb1/tmpfile.{1..16} /mnt/sdc1/tmpfile.{1..16} /mnt/sdd1/tmpfile.{1..16} /mnt/sde1/tmpfile.{1..16} /mnt/sdf1/tmpfile.{1..16} /mnt/sdg1/tmpfile.{1..16} /mnt/sdh1/tmpfile.{1..16} /mnt/sdi1/tmpfile.{1..16} /mnt/sdj1/tmpfile.{1..16} /mnt/sdk1/tmpfile.{1..16} > iozone.txt

   .. list-table::
      :widths: auto
      :width: 100%

      * - ``-s 1g``
        - Size of 1G per file
      * - ``-r`` 
        - 4m  4MB block size
      * - ``-i #``   
        - 0=write/rewrite, 1=read/re-read, 2=random-read/write
      * - ``-I``     
        - Direct-IO modern
      * - ``-t N``   
        - Number of threads (:math:`numberOfDisks * 16`)
      * - ``-F <>``  
        - list of files (the above command tests with 16 files per disk)  

MinIO Tests
~~~~~~~~~~~

> Run the three :mc:`mc support perf` tests.

Reviewer: Provide examples below, even if we have more examples on the linked page. Esp. in case we add/remove tests later. Easier to say "Perform each of the following :mc:`mc support perf` tests.

Author: Copied the examples from the mc support perf docs over.

   MinIO does not provide hosted services or hardware sales.

.. list-table::
   :widths: auto
   :width: 100%

   * - :octicon:`circle`
     - | Sufficient CPU cores to acheive performance goals for hashing (for example, for healing) and encryption
       | MinIO recommends Dual Intel® Xeon® Scalable Gold CPUs (minimum 8 cores per socket) or any CPU with AVX512 instructions

   * - :octicon:`circle`
     - | Sufficient RAM to achieve performance goals based on the number of drives and anticipated concurrent requests (see the :ref:`formula and reference table <minio-requests-per-node>`)
       | Refer to the information on :ref:`memory allocation <minio-k8s-production-considerations-memory>` for recommended RAM amounts 

   * - :octicon:`circle`
     - Four nodes or servers

   * - :octicon:`circle`
     - | SATA/SAS drives for capacity and NVMe SSDs for high-performance
       | MinIO recommends a minimum of 8 drives per server

   * - :octicon:`circle`
     - | 25GbE network for capacity 
       | 100GbE Network interface cards for high performance

.. important:: 

>    Object storage operations are primarily limited by network throughput.
>    Additional performance limits can be imposed by the storage controller and the storage devices themselves. 

Reviewer: Remove ze passives. Spitballing

```suggestion
   Object storage performance typically bottlenecks with the following resources, in order of importance:
   
   - Network (Insufficient or limited throughput)
   - Storage Controller (Old firmware, limited throughput)
   - Storage (Old firmware, slow/aged/failing drives)
   
   Prioritize validating these layers before focusing on potential compute-related performance constraints.
```

Author: 👍

   For example:

>    - 1GbE network link can support approximately 125MB/s of throughput - which may not support the full capabilities of even a single spinning disk

Reviewer: cc @eco-minio to confirm my suggested math is math.

Reviewer 2: It's a math-ish. 125 is your upper bound, realistic will be a bit lower due to tcp overhead. 100-110ish is average, under 90 indicates an issue (bad firmware, overloaded box/network, etc)

Author: Updated with 
> 1GbE network link can support up to 125MB/s of throughput with 100-110MB/s as an average, which may not support the full capabilities of even a single spinning disk

   - 10GbE network can support approximately 1.25GB/s, potentially supporting 4-6 spinning disk drives running at ~200MB/s (such as Western Digital Gold drives)
   - 25GbE network can support approximately 3.125GB/s, potentially supporting 12 disks

   The recommended minimum MinIO cluster of 4 nodes with 4 drives each (16 total disks) requires a 50GbE network to support the total potential aggregate throughput.

   MinIO takes full advantage of the modern hardware improvements such as AVX-512 SIMD acceleration, 100GbE networking, and NVMe SSDs when available.

Recommended Hardware Tests
--------------------------

MinIO Diagnostics
~~~~~~~~~~~~~~~~~

Run the following command.
If you have access to :ref:`SUBNET <minio-docs-subnet>`, you can upload the results there.

.. code-block:: shell
   :class: copyable

   mc support diag ALIAS

Drive Tests
~~~~~~~~~~~

Test each drive independently on all servers to ensure they are identical in performance.
Use the results of these OS-level tools to verify the capabilities of your storage hardware.
Record the results for later reference.

> #. Test the drive's performance during write opeartions

Reviewer: ```suggestion
#. Test the drive's performance during write operations
```

Author: You know, science has shown that the order of letters in a word is less important than the beginning and ending letters.

Fixed anyway, though. Just because.

>    This tests checks a drive's ability to write new data (uncached) to disk by creating a specified number of blocks at up to a certain number of bytes at a time to mimic how a drive would function with writing uncached data. 

Reviewer: as an afterthought - we avoid the cache because it produces inconsistent results. We want to see the actual drive performance with consistent file IO, not just what the drive can do in "burst".

Author: 👍

#. Test the drive's performance during read operations

   .. code-block::
      :class: copyable

      dd if=/mnt/driveN/testfile of=/dev/null bs=128k iflag=direct > dd-read-drive1.txt

   Replace ``driveN`` with the path for the disk you are testing.

   .. list-table::
      :widths: auto
      :width: 100%

      * - ``dd``
        - The command to copy and paste data
      * - ``if=/mnt/driveN/testfile``
        - Read from ``/mnt/driveN/testfile``; replace with the path to the file to use for testing the drive's read performance
      * - ``of=/dev/null``
        - Write to ``/dev/null``, a virtual file that does not persist after the operation completes
      * - ``bs=128k``
        - Write up to 128,000 bytes at a time
      * - ``count=80000``
        - Write up to 80000 blocks of data
      * - ``iflag=direct``
        - Use direct I/O to read and avoid data from caching
      * - ``> dd-read-drive1.txt``
        - Write the contents of the operation's output to ``dd-read-drive1.txt`` in the current working directory

   Use a sufficiently sized file that mimics the primary use case for your deployment to get accurate read test results.
   

>    Small files have a size < 1MB.
>    Large files are from 1MB — 1GB.
>    Really large files are > 1GB.

Reviewer: ```suggestion
The following guidelines may help during performance testing:

- Small Files : < 128KB
- Normal Files: <1GB
- Large Files: > 1GB
```

I realized in re-reading that our small file logic specifically kicks in at 128KB so that's probably where we want to focus testing.

Author: 👍

   :local:
   :depth: 2

Use the following checklist when planning the hardware configuration for a production, distributed MinIO deployment.

Considerations
--------------

When selecting hardware for your MinIO implementation, take into account the following factors:

- Expected amount of data in tebibytes to store at launch
- Expected growth in size of data for at least the next two years
- Number of objects by average object size
- Average retention time of data in years
- Number of sites to be deployed

Hardware Requirements and Recommendations
-----------------------------------------

.. note:: 

   See our `Reference Hardware <https://min.io/product/reference-hardware?ref-docs>`__ page for a curated selection of servers and storage components from our hardware partners.

   MinIO does not provide hosted services or hardware sales.

.. list-table::
   :widths: auto
   :width: 100%

   * - :octicon:`circle`
     - | Sufficient CPU cores to acheive performance goals for hashing (for example, for healing) and encryption

>        | MinIO recommends Dual Intel® Xeon® Scalable Gold CPUs (minimum 8 cores per socket) or any CPU with AVX512 instructions

Reviewer 2: AFAIK avx512 is a nice to have, but since a) we don't use much CPU regardless and b) the performance gains for avx512 vs avx2 are not enough to make things be mandatory to have avx512 in play. I sort of waffle on this since, no, it is not needed strictly, but on the other side, if you have an Intel processor that doesn't have 512, it might also indicate the age of the hardware could have other issues (AMD is a different story since they do not have avx512 or didn't introduce it until recently). So all that a very long winded way to ask if mandating 512 causes more confusion than it solves?

Author: We're not mandating. We're recommending.
> **MinIO recommends** .... or any CPU with AVX512 instructions.

Do you think that is sufficient, @eco-minio? Or do we just need to remove that.

Reviewer 2: I just want to mitigate the conversation of "but, I only have avx2, am I doomed?" but probably for the amount of times we have to clarify it might not be worth the effort to get verbiage around. I would stick with what you have for now and we can amend in the future if it ever was needed

Reviewer: Maybe if we trim the H1 to just "Hardware Recommendations", since we do have "Requirements and Recommendations"

That said, we're allowed to be prescriptive. We could try to address this in the lede

```
Hardware Requirements
--------------------------------

The following table provides a minimum hardware specification for production MinIO deployments. 
While MinIO can run on commodity or "budget" hardware, we strongly recommend using this table as guidance for best results in production environments.
```

   * - :octicon:`circle`
     - | Sufficient RAM to achieve performance goals based on the number of drives and anticipated concurrent requests (see the :ref:`formula and reference table <minio-requests-per-node>`)
       | Refer to the information on :ref:`memory allocation <minio-k8s-production-considerations-memory>` for recommended RAM amounts 

   * - :octicon:`circle`

>      - Four nodes or servers

Reviewer 2: Should call out that for VM's / k8s workers we need four distinct physical nodes

Author: 👍

   * - :octicon:`circle`
     - | SATA/SAS drives for capacity and NVMe SSDs for high-performance
       | MinIO recommends a minimum of 8 drives per server

   * - :octicon:`circle`
     - | 25GbE network for capacity 
       | 100GbE Network interface cards for high performance

.. important:: 

   The following areas have the greatest impact on MinIO performance, listed in order of importance:

   - Network infrastructure (insufficient or limited throughput)
   - Storage controller (old firmare, limited throughput)

>    - Storge (old firmware; slow, aged, or failing drives)

Reviewer 2: ```suggestion
   - Storage (old firmware; slow, aged, or failing drives)
```

Author: 🙃

   Prioritize upgrading these areas before focusing on compute-related performance constraints.
   
   For example:

   - 1GbE network link can support approximately 125MB/s of throughput - which may not support the full capabilities of even a single spinning disk
   - 10GbE network can support approximately 1.25GB/s, potentially supporting 4-6 spinning disk drives running at ~200MB/s (such as Western Digital Gold drives)

>    - 25GbE network can support approximately 3.125GB/s, potentially supporting 12 disks

Reviewer 2: 250MB/s spindle assumes a lot of things, 100-150 is a more realistic target for real world

Author: Giving a range of potential disk count instead.

>    The recommended minimum MinIO cluster of 4 nodes with 4 drives each (16 total disks) requires a 50GbE network to support the total potential aggregate throughput.

Reviewer 2: Up above we say we recommend 8 drives, since we are talking minimums do we want to reiterate that, yes, 4 is possible, but 8 is better?

Author: Updated.

   MinIO takes full advantage of the modern hardware improvements such as AVX-512 SIMD acceleration, 100GbE networking, and NVMe SSDs when available.

Recommended Hardware Tests
--------------------------

MinIO Diagnostics
~~~~~~~~~~~~~~~~~

Run the following command.
If you have access to :ref:`SUBNET <minio-docs-subnet>`, you can upload the results there.

.. code-block:: shell
   :class: copyable

>    mc support diag ALIAS

Reviewer 2: `--airgap` so as not to assume cluster is registered already

Author: Nice.

Drive Tests
~~~~~~~~~~~

Test each drive independently on all servers to ensure they are identical in performance.
Use the results of these OS-level tools to verify the capabilities of your storage hardware.
Record the results for later reference.

#. Test the drive's performance during write operations

   This tests checks a drive's ability to write new data (uncached) to disk by creating a specified number of blocks at up to a certain number of bytes at a time to mimic how a drive would function with writing uncached data. 
   This allows you to see the actual drive performance with consistent file I/O.
   
   .. code-block::
      :class: copyable

>       dd if=/dev/zero of=/mnt/driveN/testfile bs=128k count=80000 oflag=direct conv=fdatasync > dd-write-drive1.txt

Reviewer 2: Drive testing outside of MinIO can for the most part be replaced with `mc support {drive,net}`. If those results come back unexpected (perf way outside of expected range) then users can fall back to dd/fio but otherwise it is a heavy lift now that we have tooling included in the product.

Author: I've created sections for mc support tests, default OS tests, and 3rd party tests.

      * - ``iflag=direct``
        - Use direct I/O to read and avoid data from caching
      * - ``> dd-read-drive1.txt``
        - Write the contents of the operation's output to ``dd-read-drive1.txt`` in the current working directory

   Use a sufficiently sized file that mimics the primary use case for your deployment to get accurate read test results.
   
   The following guidelines may help during performance testing:

   - Small files: < 128KB
   - Normal files: 128KB – 1GB
   - Large files: > 1GB

   You can use the ``head`` command to create a file to use.
   The following command example creates a 10 Gigabyte file called ``testfile``.

   .. code-block:: shell
      :class: copyable

      head -c 10G </dev/urandom > testfile

   The operation returns the number of files read, total size read in bytes, the total length of time for the operation (in seconds), and the speed of the reading in bytes per second.

#. IO Controller test
   
   Use `IOzone <http://iozone.org/>`__ to test the input/output controller and all drives in combination.
   Document the performance numbers for each server in your deployment.

   .. code-block:: shell
      :class: copyable

>       iozone -s 1g -r 4m -i 0 -i 1 -i 2 -I -t 160 -F /mnt/sdb1/tmpfile.{1..16} /mnt/sdc1/tmpfile.{1..16} /mnt/sdd1/tmpfile.{1..16} /mnt/sde1/tmpfile.{1..16} /mnt/sdf1/tmpfile.{1..16} /mnt/sdg1/tmpfile.{1..16} /mnt/sdh1/tmpfile.{1..16} /mnt/sdi1/tmpfile.{1..16} /mnt/sdj1/tmpfile.{1..16} /mnt/sdk1/tmpfile.{1..16} > iozone.txt

Reviewer 2: Same comment as above for dd, iozone/fio can be tabled in liew of just mc support drive.

Author: As above, split out iozone into a "Third Part Diagnostics" section.

       | MinIO recommends Dual Intel® Xeon® Scalable Gold CPUs (minimum 8 cores per socket) or any CPU with AVX512 instructions

   * - :octicon:`circle`
     - | Sufficient RAM to achieve performance goals based on the number of drives and anticipated concurrent requests (see the :ref:`formula and reference table <minio-requests-per-node>`)
       | Refer to the information on :ref:`memory allocation <minio-k8s-production-considerations-memory>` for recommended RAM amounts 

   * - :octicon:`circle`
     - | Four nodes or servers
       | For containers or Kubernetes in virtualized environments, MinIO requires four distinct physical nodes.

   * - :octicon:`circle`
     - | SATA/SAS drives for capacity and NVMe SSDs for high-performance
       | MinIO recommends a minimum of 8 drives per server

   * - :octicon:`circle`
     - | 25GbE network for capacity 
       | 100GbE Network interface cards for high performance

.. important:: 

   The following areas have the greatest impact on MinIO performance, listed in order of importance:

   - Network infrastructure (insufficient or limited throughput)

>    - Storage controller (old firmware, limited throughput)

Reviewer: ```suggestion
   - Storage controller (old firmware; limited throughput)
```

For consistency with third bullet?

Author: Sure.

   - Storage (old firmware; slow, aged, or failing drives)

   Prioritize upgrading these areas before focusing on compute-related performance constraints.
   

>    For example:

Reviewer: Thinking about Eco's notes of 110-120MB/s being typical for a spinning disk. Lets assume 100 for easy math.

```suggestion
The following examples of network throughput constraints assume spinning disks with ~100MB/s sustained IO.
```

Author: 👍

   - 1GbE network link can support up to 125MB/s of throughput with 100-110MB/s as an average, which may not support the full capabilities of even a single spinning disk

>    - 10GbE network can support approximately 1.25GB/s, potentially supporting 4-6 spinning disk drives running at ~200MB/s (such as Western Digital Gold drives)

Reviewer: Update these for 100MB/S disks. So this line should be ~10-12 spinning disks, next line would be ~30 spinning disks.

It also means 25GbE is in fact acceptable for our minimal production setup of spinning-disks only, assuming again a baseline of 100MB/S.

Author: 👍

   - 25GbE network can support approximately 3.125GB/s, potentially supporting a range of 10 to 20 disks depending on drive throughput

>    The recommended minimum MinIO cluster of 4 nodes with 4 drives each (16 total disks) requires a 50GbE network to support the total potential aggregate throughput.

Reviewer: 25GbE instead of 50GbE by updating to @eco-minio 's reccs of 100MB/s

Author: 👍
